src/Subscription.js

```diff
@@ -89,14 +74,72 @@ export default class Subscription extends EventEmitter {
             previousMsgRef.compareTo(this.lastReceivedMsgRef[key]) === 1
     }
 
-    handleMessage(msg, isResend = false) {
+    handleBroadcastMessage(msg, verificationPromise) {
+        return this._handleMessage(msg, verificationPromise, false)
+    }
+
+    handleResentMessage(msg, verificationPromise) {
+        if (!this.resending) {
+            throw new Error(`There is no resend in progress, but received resent message ${msg.serialize()}`)
+        } else {
+            const handleMessagePromise = this._handleMessage(msg, verificationPromise, true)
+            this._lastMessageHandlerPromise = handleMessagePromise
+            return handleMessagePromise
+        }
+    }
+
+    handleResending(response) {
+        if (!this.resending) {
+            throw new Error(`There should be no resend in progress, but received ResendResponseResending message ${response.serialize()}`)
+        }
+        this.emit('resending', response)
+    }
+
+    handleResent(response) {
+        if (!this.resending) {
+            throw new Error(`There should be no resend in progress, but received ResendResponseResent message ${response.serialize()}`)
+        }
+        // Delay event emission until the last message in the resend has been handled
+        if (this._lastMessageHandlerPromise) {
+            this._lastMessageHandlerPromise.then(() => {
+                this.emit('resent', response)
+                this._finishResend()
+            })
+        } else {
+            throw new Error('Attempting to handle ResendResponseResent, but no messages have been received!')
+        }
+    }
+
+    handleNoResend(response) {
+        if (!this.resending) {
+            throw new Error(`There should be no resend in progress, but received ResendResponseNoResend message ${response.serialize()}`)
+        }
+        this.emit('no_resend', response)
+        this._finishResend()
+    }
+
+    _finishResend() {
+        this._lastMessageHandlerPromise = null
+        this.setResending(false)
+        this.checkQueue()
+    }
+
+    async _handleMessage(msg, verificationPromise, isResend = false) {
```

Comment:
The way this verificationPromise is passed around is highly abnormal, makes it difficult to verify visually whether the promise gets sequenced correctly. Also looks like it can lead to unhandled rejections being reported if the verification fails and `const valid = await verificationPromise` is never called.

I think if the verification was passed as function that would make things easier to follow and much more "normal":

```js
const valid = await verify()
```
Would require minimal change to existing code & tests.

This also means verification won't even start if the handler fails for some other reason, removing the possible unhandled rejection and also avoids doing verifications that can never be used.

Reply:
Another idea is an API like this:

```js
async isVerified(msg) {
  this._verifiedMessages = this._verifiedMessages || new WeakMap()

  const stream = this.subscribedStreams[msg.streamMessage.getStreamId()]
  const isVerifiedPromise = this._verifiedMessages.get(message) || stream.verifyStreamMessage(msg.streamMessage)
  this._verifiedMessages.set(msg, isVerifiedPromise)
  return isVerifiedPromise
}
```

However the above code won't work as-is since `Subscription` doesn't currently have access to the stream in order to call `stream.verifyStreamMessage`.

Reply:
The key goal here is that `stream.verifyStreamMessage(msg.streamMessage)` only gets called once. If we pass it wrapped into a function (`() => stream.verifyStreamMessage(msg.streamMessage)`), or pass `stream` to the handler to give `Subscription` access to it, each individual handler would verify the message separately, which we don't want.

So effectively we want to cache the result of verification, and passing in the promise which each handler can query for the result seemed like by far the simplest solution. I'm not sure why it is so unusual: it's actually very handy, although as you said it might not be very readable. We could use the `WeakMap` solution, although IMO it is much more complicated. Note that we'd need to share that `WeakMap` instance among all `Subscription` objects, it can't be an instance variable as in your example.

We could pass in a function which returns a singleton promise, which would of course be a bit silly compared to just passing in the promise. But would that make the interface more conventional?

We could also wrap the `stream.verifyStreamMessage(msg.streamMessage)` call into a more complex function that only executes `stream.verifyStreamMessage(msg.streamMessage)` on the first call, and thereafter returns a cached promise on every subsequent call, but that too would just add a function wrapping around what essentially is returning the same promise to each handler, which is what we're doing already.

Reply:
> We could also wrap the stream.verifyStreamMessage(msg.streamMessage) call into a more complex function that only executes stream.verifyStreamMessage(msg.streamMessage) on the first call

This is what I ended up doing. The code ended up relatively clean. I was able to use a ready-made package `once`. I'm glad I discovered this package: I've been missing it for a while, and have seen people roll their own.